### PR TITLE
Sentry

### PR DESCRIPTION
**Pull request by davrv93**
Happy week to everybody!
I've created a Hook for sentry.
To enable the mechanism you have to set :
$config['enable_hooks'] = TRUE;
(See config.php.example for example)

The hook is loaded after the controller finish the constructor, and executes captureLastError() function to send the information to sentry. You can try adding this line : echo $data[2]; to Home.php on line 40.

image

image

It will create a register on sentry logs

image
image

It shows the line of the error, information about the client: Browser, IP, WebView, Headers and other important things.

If you want to manually catch the exceptions you can disable hooks and use this:

try {
    $this->functionFailsForSure();
} catch (\Throwable $exception) {
    Sentry\captureException($exception);
}
I would like to know your opinion, thanks.
Blessings!

### CODE REVIEW

File: application/controllers/Home.php [hunk 1]
```diff
@@ -16,11 +16,16 @@ public function __construct()
 		$this->load->database();
 		$this->load->library( [ "ion_auth", "twig" ] );
 		$this->load->model( "product_model" );
+
```
Comment (vibby) [inline application/controllers/Home.php]: Uneeded change, isn’t it ?

Comment (davrv93) [reply application/controllers/Home.php]: Yes you're right, I have revert the line

**Comment by davrv93** [comment]
Hi @vibby  ! I have made some changes on this PR:
I do the following steps:

I have added on apache2 server an environment variable
SetEnv CI_ENV development

Then I create a folder called development inside config folder.

Inside it I created a file called config.php and inside it I put a variable called
$config['sentry'] = TRUE
and also
$config['sentry_dsn'] = dsn value

So my folders looks like this:

image

And finally the hook looks like this

image


and the bugs are being reported to sentry :+1: 

image

Thanks for the good suggestion! Lord bless you.

File: application/hooks/SentryLoader.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        |--------------------------------------------------------------------------
+        |
+        | In this function we init Sentry with the setup params that were provided
+        | by sentry.io
+        */
+        $ci = get_instance(); // CI_Loader instance
+        $ci->load->config('config');
+        $sentry = $ci->config->item('sentry');
+        $sentry_dsn = $ci->config->item('sentry_dsn');
+        if($sentry==TRUE){
```
Comment (vibby) [inline application/hooks/SentryLoader.php]: I think it is better not to attach the hook when config is at false. But maybe it is not possible to reach for config inside the hooks file …

File: application/config/development/config.example.php [hunk 3]
```diff
@@ -0,0 +1,526 @@
+<?php
```
Comment (vibby) [inline application/config/development/config.example.php]: Good idea to have an example for dev and for prod. you cold also rename ```application/config/config.example.php``` to ```application/config/config.production.example.php``` or maybe ```application/config/production/config.example.php```.

File: application/hooks/SentryLoader.php (hunk 2 above)
Comment (vibby) [reply application/hooks/SentryLoader.php]: Never mind, if you have to move all code to get config to ```hooks.php```, it will not be clearer.

**Comment by davrv93** [comment]
Hi @vibby  I have made the requested changes, also I follow the codeigniter documentation for having develop and production environments., thanks!

**Comment by vibby** [comment]
Hello @davrv93. That is great ! Can you please rebase the branch on top of master in order to be able to merge it ? Let me know if you need help on that

**Comment by vibby** [comment]
I just saw your message @davrv93. I rebase it.

**Pull request merged**